spark2/src/main/java/org/apache/iceberg/spark/source/CustomCatalogs.java

```diff
@@ -0,0 +1,102 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.spark.source;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.iceberg.CatalogUtil;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.catalog.Catalog;
+import org.apache.iceberg.catalog.Namespace;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.relocated.com.google.common.base.Splitter;
+import org.apache.iceberg.spark.SparkUtil;
+import org.apache.iceberg.util.Pair;
+import org.apache.spark.SparkConf;
+import org.apache.spark.sql.SparkSession;
+
+public final class CustomCatalogs {
+  private static final Cache<Pair<SparkSession, String>, Catalog> CATALOG_CACHE = Caffeine.newBuilder().softValues().build();
+
+  public static final String ICEBERG_DEFAULT_CATALOG = "default_catalog";
+  public static final String ICEBERG_CATALOG_PREFIX = "spark.sql.catalog";
+
+  private CustomCatalogs() {
+  }
+
+  /**
+   * Build an Iceberg {@link Catalog} to be used by this Spark source adapter.
+   *
+   * <p>
+   *   The cache is to facilitate reuse of catalogs, especially if wrapped in CachingCatalog. For non-Hive catalogs all
+   *   custom parameters passed to the catalog are considered in the cache key. Hive catalogs only cache based on
+   *   the Metastore URIs as per previous behaviour.
+   * </p>
+   *
+   * @param spark Spark Session
+   * @param name Catalog Name
+   * @return an Iceberg catalog
+   */
+  public static Catalog buildIcebergCatalog(SparkSession spark, String name) {
+    return CATALOG_CACHE.get(Pair.of(spark, name), CustomCatalogs::build);
+  }
+
+  private static Catalog build(Pair<SparkSession, String> sparkAndName) {
+    SparkSession spark = sparkAndName.first();
+    String name =  sparkAndName.second() == null ? ICEBERG_DEFAULT_CATALOG : sparkAndName.second();
+    SparkConf sparkConf = spark.sparkContext().getConf();
+    Configuration conf = spark.sessionState().newHadoopConf();
+
+    String catalogPrefix = String.format("%s.%s.", ICEBERG_CATALOG_PREFIX, name);
+    Map<String, String> options = Arrays.stream(sparkConf.getAllWithPrefix(catalogPrefix))
+        .collect(Collectors.toMap(x -> x._1, x -> x._2));
+
+    if (options.isEmpty() && !name.equals(ICEBERG_DEFAULT_CATALOG)) {
+      throw new IllegalArgumentException(String.format("Cannot instantiate catalog %s. Incorrect Parameters", name));
```

Comment:
This seems like a concern for `CatalogUtil.buildIcebergCatalog`, not here. This should create the map and pass it on to allow that method to reject the options. That way, the user gets a more specific error, like "Missing catalog implementation class or type". That can also create a default catalog if we choose to later -- though I'm skeptical we would -- without needing to change all of the callers.

Reply:
the problem arises when a eg `default.table` is passed into this method. It will end up constructing a Hive catalog w/ existing options called `default`. The Hive catalog tries to pull config from `Configuration` if not passed via the constructor. So we eat the namespace and using it in the catalog name rather than constructing a default catalog and a table identifier `default.table`.

We could:
* not allow catalog names to be specified in Spark2 sources
* Stop using `Configuration` in hive tables
* keep this the way it is

I prefer option 2 but i am worried that may have unintended consequences elsewhere.

Reply:
I don't think I quite understand you here.

In Spark, a catalog exists if `spark.sql.catalog.name` is present and is a class that can be loaded. I would expect this method to return null if that property isn't the case. For Spark 2.4, I think it would be fine to assume that the class will be `SparkCatalog.class.getName()`. If it isn't, then return null. If it is, then the catalog exists and this should load the `IcebergCatalog` that would be used by `SparkCatalog` in Spark 3.x. Does that make sense?

Reply:
To put it another way. If spark is set up to with `spark.sql.catalog.name` set then behave as if you are Spark3 and return a custom catalog. Otherwise return Hive.

Reply:
Yep! I would say that this should just respect Spark 3 settings.

Reply:
cool. I have attempted to do that here. The complication is the same method fetches default catalog and custom catalog so there is an extra check there. Can separate out if you prefer.